tsrelay/handler/get_peers.go

```diff
@@ -0,0 +1,129 @@
+package handler
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"io"
+	"net"
+	"net/http"
+	"sort"
+	"strings"
+)
+
+// getPeersResponse is a subset of ipnstate.Status
+// which contains only what the plugin needs
+// to reduce serialization size in addition
+// to some helper fields for the typescript frontend
+type getPeersResponse struct {
+	Peers          []*peerStatus
+	CurrentTailnet *currentTailnet
+	Errors         []Error `json:",omitempty"`
+}
+
+type currentTailnet struct {
+	Name            string
+	MagicDNSSuffix  string
+	MagicDNSEnabled bool
+}
+
+func (h *handler) getPeersHandler(w http.ResponseWriter, r *http.Request) {
+	s, err := h.getPeers(r.Context(), r.Body)
+	if err != nil {
+		var re RelayError
+		if errors.As(err, &re) {
+			w.WriteHeader(re.statusCode)
+			json.NewEncoder(w).Encode(re)
+			return
+		}
+		h.l.Println("error creating serve:", err)
+		http.Error(w, err.Error(), 500)
+		return
+	}
+
+	json.NewEncoder(w).Encode(s)
+}
+
+func (h *handler) getPeers(ctx context.Context, body io.Reader) (*getPeersResponse, error) {
+	if h.requiresRestart {
+		return nil, RelayError{
+			statusCode: http.StatusPreconditionFailed,
+			Errors:     []Error{{Type: FlatpakRequiresRestart}},
+		}
+	}
+
+	st, err := h.lc.Status(ctx)
+	if err != nil {
+		var oe *net.OpError
+		if errors.As(err, &oe) && oe.Op == "dial" {
+			return nil, RelayError{
+				statusCode: http.StatusServiceUnavailable,
+				Errors:     []Error{{Type: NotRunning}},
+			}
+		}
+		return nil, err
+	}
+
+	s := getPeersResponse{
+		CurrentTailnet: &currentTailnet{
```

Comment:
I found when CurrentTailnet can be nil!

```
2023-08-02 22:07:57.868 [info] [tsrelay] 2023/08/02 22:07:57 http: panic serving 127.0.0.1:52382: runtime error: invalid memory address or nil pointer dereference
2023-08-02 22:07:57.868 [info] [tsrelay] goroutine 54 [running]:
2023-08-02 22:07:57.868 [info] [tsrelay] net/http.(*conn).serve.func1()
2023-08-02 22:07:57.868 [info] [tsrelay] /Users/tyler/.cache/tailscale-go/src/net/http/server.go:1854 +0xb0
2023-08-02 22:07:57.868 [info] [tsrelay] panic({0x10257a640, 0x1027b5f50})
2023-08-02 22:07:57.868 [info] [tsrelay] /Users/tyler/.cache/tailscale-go/src/runtime/panic.go:890 +0x258
2023-08-02 22:07:57.868 [info] [tsrelay] github.com/tailscale-dev/vscode-tailscale/tsrelay/handler.(*handler).getPeers(0x140000e5778?, {0x1025d87f8?, 0x14000395110?}, {0x102233bf4?, 0x140000e5798?})
2023-08-02 22:07:57.868 [info] [tsrelay] /Users/tyler/tailscale-dev/vscode-tailscale/tsrelay/handler/get_peers.go:69 +0xb0
2023-08-02 22:07:57.868 [info] [tsrelay] github.com/tailscale-dev/vscode-tailscale/tsrelay/handler.(*handler).getPeersHandler(0x1400014a2c0, {0x1025d8520, 0x140004a21c0}, 0x0?)
2023-08-02 22:07:57.868 [info] [tsrelay] /Users/tyler/tailscale-dev/vscode-tailscale/tsrelay/handler/get_peers.go:31 +0x88
2023-08-02 22:07:57.868 [info] [tsrelay] net/http.HandlerFunc.ServeHTTP(0x102578680?, {0x1025d8520?, 0x140004a21c0?}, 0x140003a82ec?)
2023-08-02 22:07:57.868 [info] [tsrelay] /Users/tyler/.cache/tailscale-go/src/net/http/server.go:2122 +0x38
2023-08-02 22:07:57.868 [info] [tsrelay] github.com/go-chi/chi/v5.(*Mux).routeHTTP(0x140001322a0, {0x1025d8520, 0x140004a21c0}, 0x140004a8100)
2023-08-02 22:07:57.868 [info] [tsrelay] /Users/tyler/go/pkg/mod/github.com/go-chi/chi/v5@v5.0.10/mux.go:444 +0x234
2023-08-02 22:07:57.868 [info] [tsrelay] net/http.HandlerFunc.ServeHTTP(0x1025a88e0?, {0x1025d8520?, 0x140004a21c0?}, 0x1c?)
2023-08-02 22:07:57.868 [info] [tsrelay] /Users/tyler/.cache/tailscale-go/src/net/http/server.go:2122 +0x38
2023-08-02 22:07:57.868 [info] [tsrelay] github.com/tailscale-dev/vscode-tailscale/tsrelay/handler.(*handler).authMiddleware.func1({0x1025d8520, 0x140004a21c0}, 0x140004a8100)
2023-08-02 22:07:57.868 [info] [tsrelay] /Users/tyler/tailscale-dev/vscode-tailscale/tsrelay/handler/auth_middleware.go:30 +0x3b0
2023-08-02 22:07:57.868 [info] [tsrelay] net/http.HandlerFunc.ServeHTTP(0x1025d8750?, {0x1025d8520?, 0x140004a21c0?}, 0x1027b5d90?)
2023-08-02 22:07:57.868 [info] [tsrelay] /Users/tyler/.cache/tailscale-go/src/net/http/server.go:2122 +0x38
2023-08-02 22:07:57.868 [info] [tsrelay] github.com/go-chi/chi/v5.(*Mux).ServeHTTP(0x140001322a0, {0x1025d8520, 0x140004a21c0}, 0x140004a8000)
2023-08-02 22:07:57.868 [info] [tsrelay] /Users/tyler/go/pkg/mod/github.com/go-chi/chi/v5@v5.0.10/mux.go:90 +0x2e4
2023-08-02 22:07:57.868 [info] [tsrelay] net/http.serverHandler.ServeHTTP({0x14000395020?}, {0x1025d8520, 0x140004a21c0}, 0x140004a8000)
2023-08-02 22:07:57.868 [info] [tsrelay] /Users/tyler/.cache/tailscale-go/src/net/http/server.go:2936 +0x2d8
2023-08-02 22:07:57.868 [info] [tsrelay] net/http.(*conn).serve(0x140003b0f30, {0x1025d87f8, 0x1400011d560})
2023-08-02 22:07:57.868 [info] [tsrelay] /Users/tyler/.cache/tailscale-go/src/net/http/server.go:1995 +0x560
2023-08-02 22:07:57.868 [info] [tsrelay] created by net/http.(*Server).Serve
2023-08-02 22:07:57.868 [info] [tsrelay] /Users/tyler/.cache/tailscale-go/src/net/http/server.go:3089 +0x520
```

If you're logged out.

Reply:
@tylersmalley fixed 👍🏼